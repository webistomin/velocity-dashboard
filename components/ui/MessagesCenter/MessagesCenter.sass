+b('messages-center')
  +e('btn')
    width: rem(50px)
    height: rem(50px)
    max-width: rem(60px)
    display: flex
    align-items: center
    justify-content: center
    color: rgb(var(--color-gray-4))
    transition: color 0.2s ease-out

    &:hover
      color: rgb(var(--color-primary))

    +m('active')
      color: rgb(var(--color-primary))

  +e('btn-content')
    position: relative

    .icon
      width: rem(23px)
      height: rem(18px)

    +media($screen-lg)
      .icon
        width: rem(26px)
        height: rem(21px)

  +e('sidebar')
    width: 356px
    height: calc(100% - 5.1rem)
    top: rem(51px)
    right: 0

    +media($screen-sm)
      top: 8rem
      height: calc(100% - 8rem)

  +e('overlay')
    height: calc(100% - 5.1rem)
    top: rem(51px)

    +media($screen-sm)
      top: 8rem
      height: calc(100% - 8rem)

    +media($screen-lg)
      display: none
