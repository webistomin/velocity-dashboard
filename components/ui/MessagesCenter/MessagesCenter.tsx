import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseNotify from 'components/base/BaseNotify';
import BaseSidebar from 'components/base/BaseSidebar';
import BaseOverlay from 'components/base/BaseOverlay';
import ChatList from 'components/ui/Chat/ChatList';
import { IChat } from 'types/chat/chat-list';

import './MessagesCenter.sass';
import { Watch } from '~/node_modules/nuxt-property-decorator';

@Component({
  name: 'MessagesCenter',
})
export default class MessagesCenter extends VueComponent {
  isMessagesVisible: boolean = false;

  chats: IChat[] = [
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
    {
      name: 'Alexey Istomin',
      avatar: '/img/avatar.png',
      message: 'You are most welcome Kathey! You are most welcome Kathey! You are most welcome Kathey!',
      id: String(Math.random()),
      time: Date.now(),
    },
  ];

  toggleMessagesVisibility() {
    this.isMessagesVisible = !this.isMessagesVisible;
  }

  get shouldLockBody(): boolean {
    return !!(this.isMessagesVisible && window.matchMedia('(max-width: 1023px)').matches);
  }

  onClickOutside(): void {
    this.isMessagesVisible = false;
  }

  @Watch('$route')
  onRouteChanged() {
    this.isMessagesVisible = false;
  }

  render(): VNode {
    return (
      <div class='messages-center' v-scroll-lock={this.shouldLockBody} v-click-outside={this.onClickOutside}>
        <button
          class={`messages-center__btn btn ${this.isMessagesVisible ? 'messages-center__btn_active' : ''}`}
          type='button'
          onClick={this.toggleMessagesVisibility}>
          <span class='messages-center__btn-content'>
            <BaseNotify class='messages-center__status' />
            <svg-icon name='icon-mail' width={23} height={18} />
          </span>
        </button>

        <BaseSidebar isVisible={this.isMessagesVisible} class='messages-center__sidebar'>
          <ChatList chats={this.chats} />
        </BaseSidebar>

        <BaseOverlay
          class='messages-center__overlay'
          isVisible={this.isMessagesVisible}
          onClick={this.onClickOutside}
        />
      </div>
    );
  }
}
