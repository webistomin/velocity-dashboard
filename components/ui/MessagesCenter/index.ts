import MessagesCenter from './MessagesCenter';

export default MessagesCenter;
