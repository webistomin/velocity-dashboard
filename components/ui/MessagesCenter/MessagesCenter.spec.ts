import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
// @ts-ignore
import vClickOutside from 'v-click-outside';
import VScrollLock from 'v-scroll-lock';
import MessagesCenter from './MessagesCenter';

describe('MessagesCenter UI', () => {
  const localVue = createLocalVue();
  localVue.use(VueRouter);
  localVue.use(vClickOutside);
  localVue.use(VScrollLock);
  const router = new VueRouter();

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: jest.fn(() => {
        return {
          matches: true,
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
        };
      }),
      writable: true,
    });
  });

  const factory = (propsData = {}, options = {}) => {
    return shallowMount(MessagesCenter, {
      localVue,
      router,
      propsData: {
        ...propsData,
      },
      stubs: ['svg-icon'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Toggle messages visibility', () => {
    const wrapper = factory();
    wrapper.setData({
      isMessagesVisible: false,
    });
    wrapper.vm.toggleMessagesVisibility();
    expect(wrapper.vm.isMessagesVisible).toEqual(true);
  });

  it('Close modal on "Esc"', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).toHaveBeenCalled();
  });

  it('Delete global event listeners on destroy', () => {
    const mockedCloseModal = jest.fn();
    const wrapper = factory(
      {},
      {
        attachToDocument: true,
        methods: {
          toggleModalVisibilityByKeyDown: mockedCloseModal,
        },
      }
    );
    wrapper.destroy();
    wrapper.trigger('keydown.esc', { key: 'Escape' });
    expect(mockedCloseModal).not.toHaveBeenCalled();
  });

  it('Close modal on route change', async () => {
    const wrapper = factory();
    wrapper.setData({
      isMessagesVisible: true,
    });
    await wrapper.vm.$router.push({ hash: '#' });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isMessagesVisible).toEqual(false);
  });
});
