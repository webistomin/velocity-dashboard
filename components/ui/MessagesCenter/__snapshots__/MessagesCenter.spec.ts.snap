// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesCenter UI Matches snapshot 1`] = `
"<div class=\\"messages-center\\"><button type=\\"button\\" aria-label=\\"Toggle messages center\\" class=\\"messages-center__btn btn \\"><span class=\\"messages-center__btn-content\\"><vuecomponent-stub class=\\"messages-center__status\\"></vuecomponent-stub><svg-icon-stub name=\\"icon-mail\\" width=\\"23\\" height=\\"18\\"></svg-icon-stub></span></button>
  <vuecomponent-stub class=\\"messages-center__sidebar\\">
    <chatlist-stub chats=\\"[object Object]\\"></chatlist-stub>
  </vuecomponent-stub>
  <baseoverlay-stub class=\\"messages-center__overlay\\"></baseoverlay-stub>
</div>"
`;
